### PR TITLE
implement api campaign comment

### PR DESCRIPTION
**Pull request by FangHaiWu**

### CODE REVIEW

File: src/DonateHope.Core/DTOs/CampaignCommentDTOs/CampaignCommentCreateRequestDto.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+using DonateHope.Domain.Entities;
+using DonateHope.Domain.IdentityEntities;
+namespace DonateHope.Core.DTOs.CampaignCommentDTOs;
+
+public class CampaignCommentCreateRequestDto
+{
+    // public Guid Id { get; set; }
```
Comment (AnNguyenLe) [inline src/DonateHope.Core/DTOs/CampaignCommentDTOs/CampaignCommentCreateRequestDto.cs]: If we don't need it, delete it

File: src/DonateHope.Core/Services/CampaignCommentsServices/CampaignCommentUpdateService.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                updateCommentRequestDto.Id
+                );
+                 return new ProblemDetailsError("You are unauthorized to update this campaign comment.");
+        }
+
+        var updatedCampaignComment = _campaignCommentMapper.MapCampaignCommentUpdateRequestDtoToCampaignComment(
+            updateCommentRequestDto
+        );
+        // updatedCampaignComment.UserId = userId;
+        // updatedCampaignComment.CampaignId = currentCampaignComment.CampaignId;
```
Comment (AnNguyenLe) [inline src/DonateHope.Core/Services/CampaignCommentsServices/CampaignCommentUpdateService.cs]: Comment is not necessary here. Let's delete it

File: src/DonateHope.Core/Services/CampaignCommentsServices/CampaignCommentDeleteService.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                "Failed to delete campaign comment {CampaignCommentId}. Error: {ErrorMessage}",
+                campaignCommentId,
+                deletedResult.Errors.First().Message
+                );
+            return new ProblemDetailsError("Failed to delete campaign comment.");
+        }
+
+        _logger.LogInformation(
+            "Successfully deleted campaign contribution {CampaignContributionId}", campaignCommentId);
+        return _campaignCommentMapper.MapCampaignCommentToCampaignCommentDeleteDto(deletedCampaignComment);
```
Comment (AnNguyenLe) [inline src/DonateHope.Core/Services/CampaignCommentsServices/CampaignCommentDeleteService.cs:63]: For the the deletion, we don't have to return any value. We just need to know if the delete action was succeeded or not. For such case, these lines:
`deletedCampaignComment.DeletedAt = DateTime.UtcNow;
  deletedCampaignComment.DeletedBy = deletedBy;`
  is no longer necessary.

File: src/DonateHope.WebAPI/Controllers/v1/CampaignComment/CampaignCommentController.cs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            updateCampaignCommentDto,
+            parsedUserId
+        );
+
+        if (updatedResult.IsFailed)
+        {
+            return updatedResult.Errors.ToDetailedBadRequest();
+        }
+
+        return CreatedAtRoute(
```
Comment (AnNguyenLe) [inline src/DonateHope.WebAPI/Controllers/v1/CampaignComment/CampaignCommentController.cs]: This update request does not create a new resource. Using http code 204 should be enough

File: src/DonateHope.WebAPI/Controllers/v1/CampaignComment/CampaignCommentController.cs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+            campaignCommentId,
+            deletedBy,
+            reasonForDeletionRequestDto.ReasonForDeletion
+            );
+
+        if (result.IsFailed)
+        {
+            return result.Errors.ToDetailedBadRequest();
+        }
+        return result.Value;
```
Comment (AnNguyenLe) [inline src/DonateHope.WebAPI/Controllers/v1/CampaignComment/CampaignCommentController.cs]: Delete action should not return any value. Http code 200 should be enough. Also, please remove any code related to return the value if any

File: src/DonateHope.WebAPI/DonateHope.WebAPI.csproj [hunk 6]
```diff
@@ -18,6 +18,8 @@
     <PackageReference Include="Serilog.AspNetCore" Version="8.0.2" />
     <PackageReference Include="Serilog.Sinks.Seq" Version="8.0.0" />
     <PackageReference Include="Swashbuckle.AspNetCore" Version="6.4.0" />
+    <PackageReference Include="System.Security.Cryptography.Xml" Version="5.0.0" />
+    <PackageReference Include="System.Text.Encodings.Web" Version="4.5.1" />
```
Comment (AnNguyenLe) [inline src/DonateHope.WebAPI/DonateHope.WebAPI.csproj]: What is the purpose of adding new libs?

File: src/DonateHope.WebAPI/StartupExtensions/DependencyInjectionServicesExtension.cs [hunk 7]
```diff
@@ -46,24 +50,22 @@ IConfiguration configuration
...
+
+        services.TryAddScoped<ICampaignCommentsRepository, CampaignCommentsRepository>();
+        services.TryAddSingleton<CampaignCommentMapper>();
+        services.TryAddScoped<ICampaignCommentCreateService, CampaignCommentCreateService>();
+        services.TryAddScoped<ICampaignCommentRetrieveService, CampaignCommentRetrieveService>();
+        services.TryAddScoped<ICampaignCommentUpdateService, CampaignCommentUpdateService>();
+        services.TryAddScoped<ICampaignCommentDeleteService, CampaignCommentDeleteService>();
         return services;
     }
 }
```
Comment (AnNguyenLe) [inline src/DonateHope.WebAPI/StartupExtensions/DependencyInjectionServicesExtension.cs:71]: thank you for your hard work!

File: src/DonateHope.Core/Mappers/CampaignCommentMapper.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    [MapperIgnoreTarget(nameof(CampaignComment.BannedStatusNote))]
+    [MapperIgnoreTarget(nameof(CampaignComment.User))]
+    [MapperIgnoreTarget(nameof(CampaignComment.UserId))]
+    [MapperIgnoreTarget(nameof(CampaignComment.Campaign))]
+    [MapperIgnoreTarget(nameof(CampaignComment.CampaignId))]
+    public partial CampaignComment MapCampaignCommentUpdateRequestDtoToCampaignComment(CampaignCommentUpdateRequestDto dto);
+
+    [MapperIgnoreSource(nameof(CampaignComment.UpdatedBy))]
+    [MapperIgnoreSource(nameof(CampaignComment.UpdatedAt))]
+    public partial CampaignCommentDeleteDto MapCampaignCommentToCampaignCommentDeleteDto(CampaignComment dto);
```
Comment (AnNguyenLe) [inline src/DonateHope.Core/Mappers/CampaignCommentMapper.cs:43]: Mapper for deletion is not needed. As at the end of the action we just want to know if the delete action succeeded or not

**Review by AnNguyenLe — CHANGES_REQUESTED** [review]

File: src/DonateHope.Core/DTOs/CampaignCommentDTOs/CampaignCommentCreateRequestDto.cs (hunk 1 above)
Comment (FangHaiWu) [reply src/DonateHope.Core/DTOs/CampaignCommentDTOs/CampaignCommentCreateRequestDto.cs]: Removed it

File: src/DonateHope.Core/Services/CampaignCommentsServices/CampaignCommentUpdateService.cs (hunk 2 above)
Comment (FangHaiWu) [reply src/DonateHope.Core/Services/CampaignCommentsServices/CampaignCommentUpdateService.cs]: Removed it

File: src/DonateHope.Core/Services/CampaignCommentsServices/CampaignCommentDeleteService.cs (hunk 3 above)
Comment (FangHaiWu) [reply src/DonateHope.Core/Services/CampaignCommentsServices/CampaignCommentDeleteService.cs:63]: Fixed

File: src/DonateHope.WebAPI/Controllers/v1/CampaignComment/CampaignCommentController.cs (hunk 4 above)
Comment (FangHaiWu) [reply src/DonateHope.WebAPI/Controllers/v1/CampaignComment/CampaignCommentController.cs]: Revised code.

File: src/DonateHope.WebAPI/Controllers/v1/CampaignComment/CampaignCommentController.cs (hunk 5 above)
Comment (FangHaiWu) [reply src/DonateHope.WebAPI/Controllers/v1/CampaignComment/CampaignCommentController.cs]: Fixed "return result.Value" to "return Ok()"

File: src/DonateHope.Core/Services/CampaignCommentsServices/CampaignCommentDeleteService.cs (hunk 3 above)
Comment (FangHaiWu) [reply src/DonateHope.Core/Services/CampaignCommentsServices/CampaignCommentDeleteService.cs:63]: deletedCampaignComment.DeletedAt = DateTime.UtcNow; deletedCampaignComment.DeletedBy = deletedBy;
De luu lai thoi gian xoa

**Review by FangHaiWu — COMMENTED** [review]
Nhờ Ân check lại

**Review by AnNguyenLe — APPROVED** [review]

**Pull request merged**